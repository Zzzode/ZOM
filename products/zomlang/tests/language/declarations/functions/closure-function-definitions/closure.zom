// RUN: %zomc compile --dump-ast %s | %FileCheck %s

// Test closure function definition with typed parameters
let closure = fun (n: i32, s: str) -> str {
  "1234";
}


// CHECK: {
// CHECK-NEXT:   "node": "SourceFile",
// CHECK-NEXT:   "fileName": "{{.*closure.zom}}",
// CHECK-NEXT:   "children": [
// CHECK-NEXT:     {
// CHECK-NEXT:       "node": "VariableDeclaration",
// CHECK-NEXT:       "bindings": [
// CHECK-NEXT:         {
// CHECK-NEXT:           "node": "BindingElement",
// CHECK-NEXT:           "name": "closure",
// CHECK-NEXT:           "initializer": 
// CHECK-NEXT:           {
// CHECK-NEXT:             "node": "FunctionExpression",
// CHECK-NEXT:             "parameters": [
// CHECK-NEXT:               {
// CHECK-NEXT:                 "node": "BindingElement",
// CHECK-NEXT:                 "name": "n",
// CHECK-NEXT:                 "varType": 
// CHECK-NEXT:                 {
// CHECK-NEXT:                   "node": "PredefinedType",
// CHECK-NEXT:                   "name": "i32"
// CHECK-NEXT:                 }
// CHECK-NEXT:               },
// CHECK-NEXT:               {
// CHECK-NEXT:                 "node": "BindingElement",
// CHECK-NEXT:                 "name": "s",
// CHECK-NEXT:                 "varType": 
// CHECK-NEXT:                 {
// CHECK-NEXT:                   "node": "PredefinedType",
// CHECK-NEXT:                   "name": "str"
// CHECK-NEXT:                 }
// CHECK-NEXT:               }
// CHECK-NEXT:             ],
// CHECK-NEXT:             "returnType":             {
// CHECK-NEXT:               "node": "ReturnType",
// CHECK-NEXT:               "returnType": 
// CHECK-NEXT:               {
// CHECK-NEXT:                 "node": "PredefinedType",
// CHECK-NEXT:                 "name": "str"
// CHECK-NEXT:               }
// CHECK-NEXT:             },
// CHECK-NEXT:             "body": 
// CHECK-NEXT:             {
// CHECK-NEXT:               "node": "BlockStatement",
// CHECK-NEXT:               "statements": [
// CHECK-NEXT:                 {
// CHECK-NEXT:                   "node": "ExpressionStatement",
// CHECK-NEXT:                   "expression": 
// CHECK-NEXT:                   {
// CHECK-NEXT:                     "node": "StringLiteral",
// CHECK-NEXT:                     "value": "\"1234\""
// CHECK-NEXT:                   }
// CHECK-NEXT:                 }
// CHECK-NEXT:               ]
// CHECK-NEXT:             }
// CHECK-NEXT:           }
// CHECK-NEXT:         }
// CHECK-NEXT:       ]
// CHECK-NEXT:     }
// CHECK-NEXT:   ]
// CHECK-NEXT: }
