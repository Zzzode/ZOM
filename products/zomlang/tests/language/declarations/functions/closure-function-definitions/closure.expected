{
  "type": "SourceFile",
  "fileName": "products/zomlang/tests/language/declarations/functions/closure-function-definitions/closure.zom",
  "children": [
    {
      "type": "VariableDeclaration",
      "bindings": [
        {
          "type": "BindingElement",
          "name": "closure",
          "initializer": 
          {
            "type": "FunctionExpression",
            "parameters": [
              {
                "type": "BindingElement",
                "name": "n",
                "varType": "Type"
              },
              {
                "type": "BindingElement",
                "name": "s",
                "varType": "Type"
              }
            ],
            "returnType": {"type": "Type"},
            "body": 
            {
              "type": "BlockStatement",
              "statements": [
                {
                  "type": "ExpressionStatement",
                  "expression": 
                  {
                    "type": "StringLiteral",
                    "value": "\"1234\""
                  }
                }
              ]
            }
          }
        }
      ]
    }
  ]
}